### PR TITLE
fix(lsp): don't reapply already applied hints

### PR DESCRIPTION
**Pull request by dchinmay2**
the decoration provider would clear the whole buffer then redraw all the hints every time the window was redrawn, now it stores the last applied version for a line and only clears and redraws the line if the buffer version has changed

### CODE REVIEW

**Comment by p00f** [comment]
@luukvbaal does this fix the infinite loop problem?

File: runtime/lua/vim/lsp/_inlay_hint.lua [hunk 1]
```diff
@@ -238,9 +239,12 @@ api.nvim_set_decoration_provider(namespace, {
       return
     end
     local hints_by_client = bufstate.client_hint
-    api.nvim_buf_clear_namespace(bufnr, namespace, 0, -1)
 
     for lnum = topline, botline do
+      if bufstate.applied[lnum] == bufstate.version then
+        goto continue
+      end
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/_inlay_hint.lua:246]: goto is afaik only available with luajit not with plain lua 5.1 and neovim targets plain lua 5.1 as well

**Comment by p00f** [comment]
It significantly reduces CPU usage when scrolling a large file line-by-line using `C-e` and matches usage with inlay hints disabled

File: runtime/lua/vim/lsp/_inlay_hint.lua (hunk 1 above)
Comment (p00f) [reply runtime/lua/vim/lsp/_inlay_hint.lua:246]: What do I do instead?

**Comment by luukvbaal** [comment]
> @luukvbaal does this fix the infinite loop problem?

It does👍🏻

**Review by mfussenegger — APPROVED** [review]

**Review by justinmk — APPROVED** [review]

**Pull request merged**